File: src/main/java/com/viral32111/discordrelay/discord/Gateway.java
package com.viral32111.discordrelay.discord;

import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.viral32111.discordrelay.Config;
import com.viral32111.discordrelay.Utilities;
import com.viral32111.discordrelay.discord.types.OperationCode;

import java.net.URI;
import java.net.http.WebSocket;
import java.time.Duration;
import java.util.ArrayList;
import java.util.concurrent.*;

// Client for the Discord Gateway
public class Gateway implements WebSocket.Listener {

	// The underlying websocket connection, used for stopping
	private static WebSocket webSocket = null;

	// Future that completes when the websocket connection closes, used for reconnecting depending on the boolean
	private static CompletableFuture<Boolean> connectionClosedFuture = null;

	// Compression should not be used, it is not implemented
	private static final boolean USE_COMPRESSION = false;

	// Holds a list of  received message fragments for forming the final message
	private final ArrayList<CharSequence> messageFragments = new ArrayList<>();

	// The current payload sequence number
	private Integer sequenceNumber = null;

	// The futures for heartbeating
	private CompletableFuture<?> heartbeatFuture = null;
	private CompletableFuture<?> heartbeatAcknowledgementFuture = null;

	// Starts the initial connection to the gateway
	public static void Start() {

		// Fetch the gateway URL
		API.Request( "GET", "gateway/bot", null, null ).thenAccept( ( JsonObject response ) -> {

			// Add connection properties to the returned URL
			URI url = URI.create( String.format( "%s/?v=%s&encoding=json&compress=%b", response.get( "url" ).getAsString(), Config.Get( "discord.api.version", null ), USE_COMPRESSION ) );

			// Begin the first connection to the above URL
			Connect( url, new Gateway() );

		} );

	}

	// Closes the websocket connection
	public static void Stop() throws Exception {

		// Do not continue if there is no current websocket connection
		if ( webSocket == null ) throw new Exception( "Underlying websocket connection not initialised" );

		// Send the close frame
		// NOTE: The 1000 close code is checked on response to prevennt reconnection
		webSocket.sendClose( 1000, "Goodbye" );

	}

	// Connects to a websocket server, and reconnects whenever the close future completes
	private static void Connect( URI url, Gateway listener ) {

		// Create the future for closing
		connectionClosedFuture = new CompletableFuture<>();

		// Asynchronously connect to the gateway, using the provided URL and options from the configuration
		CompletableFuture<WebSocket> webSocketConnectFuture = Utilities.HTTP_CLIENT.newWebSocketBuilder()

			// Additional headers in the upgrade request
			.header( "User-Agent", Config.Get( "http.user-agent", null ) )
			.header( "From", Config.Get( "http.from", null ) )

			// Upgrade request timeout to the server
			.connectTimeout( Duration.ofSeconds( Config.Get( "http.timeout" ) )  )

			// URL and listener instance of this class
			.buildAsync( url, listener );

		// Update the websocket property with the new websocket instance
		webSocketConnectFuture.thenAccept( ( ws ) -> webSocket = ws );

		// Reconnect to the same URL with the same listener instance whenever the future completes
		connectionClosedFuture.thenAccept( ( shouldReconnect ) -> {
			if ( shouldReconnect ) Connect( url, listener );
		} );

	}

	// Check if the websocket is currently open
	private boolean IsConnected( WebSocket webSocket ) {
		return !( webSocket.isInputClosed() || webSocket.isOutputClosed() );
	}

	// Sends a heartbeat to the gateway
	private void SendHeartbeat( WebSocket webSocket ) {

		// Create a fresh future to complete when the heartbeat is acknowledged
		heartbeatAcknowledgementFuture = new CompletableFuture<>();

		try {

			// Create the heartbeat payload
			JsonObject heartbeatPayload = new JsonObject();
			heartbeatPayload.addProperty( "op", OperationCode.Heartbeat );
			heartbeatPayload.addProperty( "d", this.sequenceNumber );

			// Send the heartbeat payload
			webSocket.sendText( heartbeatPayload.toString(), true );

		// Reconnect if any errors occur
		} catch ( Exception exception ) {
			Utilities.Error( exception );
			webSocket.sendClose( 1002, "Internal error" );
		}

		// Wait for the heartbeat acknowledgement...
		try {
			heartbeatAcknowledgementFuture.orTimeout( 5, TimeUnit.SECONDS ).get();

		// Reconnect if no acknowledgement was received
		} catch ( ExecutionException exception ) {
			Utilities.Error( "Heartbeat acknowledgement timeout: '{}'.", exception.getMessage() );
			webSocket.sendClose( 1002, "Never received heartbeat acknowledgement" );

		// Warning if the future was interrupted
		} catch ( InterruptedException exception ) {
			Utilities.Warn( "Heartbeat acknowledgement interrupted: '{}'.", exception.getMessage() );

		// Warning if the future was cancelled
		} catch ( CancellationException exception ) {
			Utilities.Warn( "Heartbeat acknowledgement cancelled: '{}'.", exception.getMessage() );

		// Error & reconnect if any other errors occur
		} catch ( Exception exception ) {
			Utilities.Error( exception );
			webSocket.sendClose( 1002, "Internal error" );
		}

	}

	// Periodically heartbeats in the background on a specified interval
	private void StartHeartbeating( WebSocket webSocket, int interval ) {

		// Display a message in the console
		Utilities.Log( "Started heartbeating every {} seconds.", interval / 1000.0 );

		// Used to check if the first heartbeat has been sent
		boolean sentInitialBeat = false;

		// Run until the websocket connection is closed...
		while ( IsConnected( webSocket ) && !heartbeatFuture.isCancelled() ) {

			try {

				// Wait a random time for the first heartbeat
				if ( !sentInitialBeat ) {
					long initialInterval = Math.round( interval * Math.random() );

					//noinspection BusyWait
					Thread.sleep( initialInterval );

					// Update variable to indicate it has been sent
					sentInitialBeat = true;

				// Wait the usual time for normal heartbeats
				} else {
					//noinspection BusyWait
					Thread.sleep( interval );
				}

			// Warn if the sleep is interrupted
			} catch ( InterruptedException exception ) {
				Utilities.Warn( "Heartbeating interrupted: '%s'", exception.getMessage() );

			// Warn if the sleep is cancelled
			} catch ( CancellationException exception ) {
				Utilities.Warn( "Heartbeating cancelled: '{}'.", exception.getMessage() );

			// Error & reconnect if any other errors occur
			} catch ( Exception exception ) {
				Utilities.Error( exception );
				webSocket.sendClose( 1002, "Internal error" );
			}

			// Check if this loop should still be running (something may have happened during the sleep)
			if ( !IsConnected( webSocket ) || heartbeatFuture.isCancelled() ) break;

			// Send a heartbeat
			this.SendHeartbeat( webSocket );

		}

		// Display a message in the console
		Utilities.Log( "Finished heartbeating." );

	}

	// Runs when the websocket connection opens
	@Override
	public void onOpen( WebSocket webSocket ) {

		// Display a message in the console
		Utilities.Log( "Gateway connection opened." );

		// Run default action
		WebSocket.Listener.super.onOpen( webSocket );

	}

	// Runs when the websocket connection closes
	@Override
	public CompletionStage<?> onClose( WebSocket webSocket, int code, String reason ) {

		// Display a message in the console
		Utilities.Log( "Gateway connection closed. ({}, '{}').", code, reason );

		try {

			// Stop the heartbeating futures
			if ( this.heartbeatFuture != null ) this.heartbeatFuture.cancel( true );
			if ( this.heartbeatAcknowledgementFuture != null ) this.heartbeatAcknowledgementFuture.cancel( true );

			// Wait for the heartbeating futures to end
			if ( this.heartbeatFuture != null ) this.heartbeatFuture.join();
			if ( this.heartbeatAcknowledgementFuture != null ) this.heartbeatAcknowledgementFuture.join();

		// Warn if the waiting is cancelled
		} catch ( CancellationException exception ) {
			Utilities.Warn( "Cancel heartbeating cancelled: '{}'", exception.getMessage() );

		// Warn if the waiting completes with an exception
		} catch ( CompletionException exception ) {
			Utilities.Warn( "Cancel heartbeating exception: '{}'", exception.getMessage() );

		// Error if any other error occures
		// NOTE: Cannot send close frame for reconnect because we are already in the close event
		} catch ( Exception exception ) {
			Utilities.Error( exception );
		}

		// Cleanup for the next run
		this.messageFragments.clear();
		this.sequenceNumber = null;
		this.heartbeatFuture = null;
		this.heartbeatAcknowledgementFuture = null;

		// Complete the future to indicate the connection is now closed
		// NOTE: This causes a reconnect if the provided code is anything other than 1000
		connectionClosedFuture.complete( code != 1000 );

		// Return default action
		return WebSocket.Listener.super.onClose( webSocket, code, reason );

	}

	// Runs when the a message is received from the websocket server
	@Override
	public CompletionStage<?> onText( WebSocket webSocket, CharSequence messageFragment, boolean isLastFragment ) {

		// Add this fragment to the array
		messageFragments.add( messageFragment );

		// If this is the last message fragment, then we have the entire message!
		if ( isLastFragment ) {

			// Join the message fragments together, and clear for the next run
			String message = String.join( "", messageFragments );
			messageFragments.clear();

			// Parse the message as JSON
			JsonObject payload = JsonParser.parseString( message ).getAsJsonObject();

			// Error & reconnect if an operation code was not included
			if ( payload.get( "op" ).isJsonNull() ) {
				Utilities.Error( "Gateway payload operation code is invalid." );
				webSocket.sendClose( 1002, "Received invalid operation code" );
				return null;
			}

			// Update the sequence number if one is included
			if ( payload.has( "s" ) && !payload.get( "s" ).isJsonNull() ) this.sequenceNumber = payload.get( "s" ).getAsInt();

			// Get the operation code from the payload
			int operationCode = payload.get( "op" ).getAsInt();

			// If this is the initial welcome message...
			if ( operationCode == OperationCode.Hello && ( payload.has( "d" ) && !payload.get( "d" ).isJsonNull() ) ) {

				// Start heartbeating using the provided interval
				int interval = payload.getAsJsonObject( "d" ).get( "heartbeat_interval" ).getAsInt();
				this.heartbeatFuture = CompletableFuture.runAsync( () -> this.StartHeartbeating( webSocket, interval ) );

				// Create the identify payload
				JsonObject identifyProperties = new JsonObject();
				identifyProperties.addProperty( "os", System.getProperty( "os.name" ) );
				identifyProperties.addProperty( "browser", Config.Get( "discord.library", null ) );
				identifyProperties.addProperty( "device", Config.Get( "discord.library", null ) );

				JsonObject identifyData = new JsonObject();
				identifyData.addProperty( "token", Config.Get( "discord.token", null ) );
				identifyData.addProperty( "intents", 1 << 9 | 1 << 15 ); // GUILD_MESSAGES and MESSAGE_CONTENT
				identifyData.addProperty( "large_threshold", 250 );
				identifyData.addProperty( "compress", USE_COMPRESSION );
				identifyData.add( "properties", identifyProperties );

				JsonObject identifyPayload = new JsonObject();
				identifyPayload.addProperty( "op", OperationCode.Identify );
				identifyPayload.add( "d", identifyData );

				// Send the identify payload
				webSocket.sendText( identifyPayload.toString(), true );

			// When a heartbeat acknowledgement is received...
			} else if ( operationCode == OperationCode.HeartbeatAcknowledgement ) {

				// Error & reconnect if we were not ready for this heartbeat acknowledgement
				if ( this.heartbeatAcknowledgementFuture == null ) {
					Utilities.Error( "Not ready for heartbeat acknowledgement" );
					webSocket.sendClose( 1002, "Received heartbeat acknowledgement too early" );
					return null;
				}

				// Complete the heartbeat acknowledgement future
				this.heartbeatAcknowledgementFuture.complete( null );

			// Send a heartbeat if the gateway requests it
			} else if ( operationCode == OperationCode.Heartbeat ) {
				this.SendHeartbeat( webSocket );

			// Error & reconnect if our session is invalid
			} else if ( operationCode == OperationCode.InvalidSession ) {
				Utilities.Error( "The session is invalid." );
				webSocket.sendClose( 1002, "Session reported as invalid" );

			// Reconnect if the gateway requests it
			} else if ( operationCode == OperationCode.Reconnect ) {
				Utilities.Log( "Gateway requested reconnect." );
				webSocket.sendClose( 1000, "Reconnect requested" );

			// If this is an event dispatch and the event name & data are present...
			} else if ( operationCode == OperationCode.Dispatch && ( payload.has( "t" ) && payload.get( "t" ).isJsonPrimitive() ) && ( payload.has( "d" ) && payload.get( "d" ).isJsonObject() ) ) {

				// Store the event name & data for easy access
				String eventName = payload.get( "t" ).getAsString();
				JsonObject eventData = payload.getAsJsonObject( "d" );

				// When we finish loading....
				if ( eventName.equals( "READY" ) ) {

					// Store the bot user
					JsonObject user = eventData.getAsJsonObject( "user" );
					String userName = user.get( "username" ).getAsString();
					Integer userTag = user.get( "discriminator" ).getAsInt();
					String userId = user.get( "id" ).getAsString();

					// Display a message in the console
					Utilities.Log( "Ready as '{}#{}' ({})", userName, userTag, userId );

				// When a new message is sent...
				} else if ( eventName.equals( "MESSAGE_CREATE" ) ) {

					// Store the message details
					String messageContent = eventData.get( "content" ).getAsString();
					String messageId = eventData.get( "id" ).getAsString();
					boolean isWebhook = eventData.has( "webhook_id" );
					String channelId = eventData.get( "channel_id" ).getAsString();

					// If this is not a webhook message...
					if ( !isWebhook ) {

						// Store the sender details
						JsonObject messageAuthor = eventData.getAsJsonObject( "author" );
						String authorName = messageAuthor.get( "username" ).getAsString();
						Integer authorTag = messageAuthor.get( "discriminator" ).getAsInt();
						String authorId = messageAuthor.get( "id" ).getAsString();
						boolean authorIsBot = ( messageAuthor.has( "bot" ) && messageAuthor.get( "bot" ).getAsBoolean() );
						boolean authorIsSystem = ( messageAuthor.has( "system" ) && messageAuthor.get( "system" ).getAsBoolean() );

						// If this is in the configured relay channel, the message has content, and the author is not a bot...
						if ( channelId.equals( Config.Get( "discord.channel.relay", null ) ) && messageContent.length() > 0 && !authorIsBot && !authorIsSystem ) {

							// Display a message in the console
							Utilities.Log( "Relaying Discord message '{}' ({}) from user '{}#{}' ({}).", messageContent, messageId, authorName, authorTag, authorId );

							// Show the message in chat to all players
							try {
								Utilities.BroadcastDiscordMessage( authorName, messageContent );

							// Error & reconnect if any other errors occur
							} catch ( Exception exception ) {
								Utilities.Error( exception );
								webSocket.sendClose( 1002, "Internal error" );
							}

						}

					}

				// Warn if we have not handled this event
				} else {
					Utilities.Warn( "Unrecognised dispatch event: '{}'.", eventName );
				}

			// Warn if we have not handled this operation code
			} else {
				Utilities.Warn( "Unrecognised gateway operation code: '{}'.", operationCode );
			}

		}

		// Return default action
		return WebSocket.Listener.super.onText( webSocket, messageFragment, isLastFragment );

	}

	// Runs when an error occurs on the websocket
	// NOTE: The websocket has already disconnected when this is called
	@Override
	public void onError( WebSocket webSocket, Throwable exception ) {

		// Display a message in the console
		Utilities.Error( "Gateway connection error: '{}'.", exception.getMessage() );

		// Run default action
		WebSocket.Listener.super.onError( webSocket, exception );

	}

}

File: src/main/kotlin/com/viral32111/discordrelay/discord/Gateway.kt
package com.viral32111.discordrelay.discord

import com.viral32111.discordrelay.DiscordRelay
import com.viral32111.discordrelay.HTTP
import com.viral32111.discordrelay.JSON
import com.viral32111.discordrelay.WebSocketCloseCode
import com.viral32111.discordrelay.config.Configuration
import com.viral32111.discordrelay.discord.data.Gateway
import com.viral32111.discordrelay.discord.data.Guild
import com.viral32111.discordrelay.helper.Version
import kotlinx.coroutines.*
import kotlinx.coroutines.future.await
import kotlinx.coroutines.time.withTimeout
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.decodeFromJsonElement
import kotlinx.serialization.json.encodeToJsonElement
import net.minecraft.server.PlayerManager
import net.minecraft.text.Style
import net.minecraft.text.Text
import net.minecraft.text.TextColor
import net.minecraft.util.Formatting
import java.io.IOException
import java.net.URI
import java.net.http.WebSocket
import java.time.Duration
import java.util.concurrent.CompletableFuture
import java.util.concurrent.CompletionStage
import kotlin.math.pow
import kotlin.random.Random

class Gateway( private val configuration: Configuration, private val playerManager: PlayerManager ) {

	// Dedicated scope for our coroutines to run on
	private val coroutineScope = CoroutineScope( Dispatchers.IO )

	// The underlying WebSocket connection
	private var webSocket: WebSocket? = null

	// Future that completes when the underlying WebSocket receives a close event
	private var connectionClosureConfirmation: CompletableDeferred<ClosureConfirmation>? = null

	// Job & completable for heartbeating in the background
	private var heartbeatJob: Job? = null
	private var heartbeatAcknowledgementConfirmation: CompletableDeferred<Unit>? = null

	// Last known sequence number for heartbeating & session resuming
	private var sequenceNumber: Int? = null

	// Used for session resuming
	private var sessionIdentifier: String? = null
	private var resumeBaseUrl: String? = null
	private var shouldResume = false

	// Number of times we've reconnected
	private var reconnectCount = 0

	// The bot ID & server roles, set during ready process
	private var myIdentifier: String? = null
	private var serverRoles: Map<String, Guild.Role>? = null

	/**
	 * Opens a WebSocket connection to the given URL, closing any existing connections beforehand.
	 * Ideally wait for closure confirmation after calling this.
	 * @param baseUrl The base URL of the WebSocket. Should not include any parameters.
	 * @param version The gateway version to include in the parameters.
	 * @return The underlying WebSocket.
	 * @see awaitClosure
	 */
	suspend fun open( baseUrl: String, version: Int = configuration.discord.api.version ) {
		val url = URI.create( "$baseUrl?v=$version&encoding=json" )

		if ( webSocket != null ) {
			DiscordRelay.LOGGER.debug( "Closing existing WebSocket connection..." )
			close( WebSocketCloseCode.GoingAway, "Closing existing connection." )
		}

		DiscordRelay.LOGGER.debug( "Opening new WebSocket connection to '$url'..." )
		connectionClosureConfirmation = CompletableDeferred()
		webSocket = HTTP.startWebSocketConnection( url, configuration.http.timeoutSeconds, Listener() )
	}

	/**
	 * Closes the underlying WebSocket connection.
	 * @param code The WebSocket close code. https://www.rfc-editor.org/rfc/rfc6455.html#section-7.4.1
	 * @param reason The human-readable reason for closing the connection.
	 */
	suspend fun close( code: Int = WebSocketCloseCode.Normal, reason: String = "Unknown.", isServerStopping: Boolean = false ) {
		try {
			DiscordRelay.LOGGER.debug( "Closing WebSocket connection with code $code & reason '$reason' (Server Stopping: $isServerStopping)..." )
			webSocket?.sendClose( code, reason )?.await()
		} catch ( exception: IOException ) {
			DiscordRelay.LOGGER.error( "Cannot close WebSocket connection! (${ exception.message })" )
		}

		DiscordRelay.LOGGER.debug( "Confirming closure..." )
		connectionClosureConfirmation?.complete( ClosureConfirmation( code, reason, isServerStopping ) )
	}

	/**
	 * Waits for a confirmation of WebSocket closure.
	 * @return The WebSocket close code, if a connection was active.
	 */
	suspend fun awaitClosure(): ClosureConfirmation? {
		DiscordRelay.LOGGER.debug( "Waiting for closure confirmation..." )
		val confirmation = connectionClosureConfirmation?.await()
		DiscordRelay.LOGGER.debug( "Closure confirmed with code ${ confirmation?.closeCode }." )
		return confirmation
	}

	// Starts heartbeating in the background
	private fun startHeartbeating( webSocket: WebSocket, interval: Long ) {
		if ( heartbeatJob != null ) {
			DiscordRelay.LOGGER.debug( "Cancelling existing background heartbeating job..." )
			heartbeatJob?.cancel()
		}

		DiscordRelay.LOGGER.debug( "Starting new background heartbeating job..." )
		heartbeatJob = coroutineScope.launch {
			heartbeatLoop( webSocket, interval )
		}
	}

	// Sends heartbeats on an interval - https://discord.com/developers/docs/topics/gateway#sending-heartbeats
	private suspend fun heartbeatLoop( webSocket: WebSocket, regularInterval: Long ) {
		val initialInterval = ( regularInterval * Random.nextFloat() ).toLong()
		DiscordRelay.LOGGER.debug( "Waiting $initialInterval milliseconds for the initial heartbeat..." )
		delay( initialInterval )

		sendHeartbeat( webSocket, this.sequenceNumber )

		while ( !webSocket.isOutputClosed ) {
			DiscordRelay.LOGGER.debug( "Waiting $regularInterval milliseconds for the next heartbeat..." )
			delay( regularInterval )

			sendHeartbeat( webSocket, this.sequenceNumber )
		}
	}

	// Sends a heartbeat event to Discord - https://discord.com/developers/docs/topics/gateway-events#heartbeat
	private suspend fun sendHeartbeat( webSocket: WebSocket, sequenceNumber: Int? ) {
		heartbeatAcknowledgementConfirmation = CompletableDeferred()

		DiscordRelay.LOGGER.debug( "Sending heartbeat at sequence number $sequenceNumber..." )
		sendEvent( webSocket, Gateway.Event.OperationCode.Heartbeat, JsonPrimitive( sequenceNumber ) )

		try {
			withTimeout( Duration.ofSeconds( configuration.discord.gateway.heartbeatTimeoutSeconds ) ) {
				DiscordRelay.LOGGER.debug( "Waiting for heartbeat acknowledgement..." )
				heartbeatAcknowledgementConfirmation?.await()
				DiscordRelay.LOGGER.debug( "Received heartbeat acknowledgement!" )
			}
		} catch ( exception: TimeoutCancellationException ) {
			DiscordRelay.LOGGER.warn( "Timed out while waiting for a heartbeat acknowledgement!" )

			DiscordRelay.LOGGER.debug( "Closing WebSocket connection..." )
			close( WebSocketCloseCode.ProtocolError, "No heartbeat acknowledgement." )
		}
	}

	// Sends an identify event to Discord - https://discord.com/developers/docs/topics/gateway#identifying
	private suspend fun sendIdentify( webSocket: WebSocket ) {
		val libraryName = "viral32111's discord relay/${ Version.discordRelay() } (https://github.com/viral32111/discord-relay)"

		// Just to be sure
		this.sessionIdentifier = null
		this.resumeBaseUrl = null
		DiscordRelay.LOGGER.debug( "Reset session identifier & resume base URL." )

		DiscordRelay.LOGGER.debug( "Sending identify..." )
		sendEvent( webSocket, Gateway.Event.OperationCode.Identify, JSON.encodeToJsonElement( Gateway.Event.Data.Identify(
			applicationToken = configuration.discord.application.token,
			intents = Gateway.Event.Data.Identify.Intents.Guilds or
					Gateway.Event.Data.Identify.Intents.GuildMessages or
					Gateway.Event.Data.Identify.Intents.MessageContent,
			connectionProperties = Gateway.Event.Data.Identify.ConnectionProperties(
				operatingSystemName = configuration.http.userAgentPrefix,
				browserName = libraryName,
				deviceName = libraryName
			)
		) ) )
	}

	// Sends a resume event to Discord - https://discord.com/developers/docs/topics/gateway#resuming
	private suspend fun sendResume( webSocket: WebSocket, sessionIdentifier: String, sequenceNumber: Int ) {
		DiscordRelay.LOGGER.debug( "Sending resume for session '$sessionIdentifier'..." )
		sendEvent( webSocket, Gateway.Event.OperationCode.Resume, JSON.encodeToJsonElement( Gateway.Event.Data.Resume(
			applicationToken = configuration.discord.application.token,
			sessionIdentifier = sessionIdentifier,
			sequenceNumber = sequenceNumber
		) ) )
	}

	// Sends an event to Discord - https://discord.com/developers/docs/topics/gateway-events#send-events
	private suspend fun sendEvent( webSocket: WebSocket, operationCode: Int, data: JsonElement? ) {
		if ( webSocket.isOutputClosed ) {
			DiscordRelay.LOGGER.warn( "WebSocket output closed when attempting to send data?!" )
			close( WebSocketCloseCode.GoingAway, "WebSocket output closed." )
			return
		}

		val jsonPayload = JSON.encodeToString( Gateway.Event(
			operationCode = operationCode,
			data = data
		) )

		DiscordRelay.LOGGER.debug( "Sending JSON payload '$jsonPayload' over WebSocket..." )
		webSocket.sendText( jsonPayload, true )?.await()
	}

	// https://discord.com/developers/docs/topics/gateway-events#hello
	private fun handleHelloEvent( webSocket: WebSocket, data: Gateway.Event.Data.Hello ) {
		val sessionIdentifier = this.sessionIdentifier
		val sequenceNumber = this.sequenceNumber

		startHeartbeating( webSocket, data.heartbeatInterval )

		coroutineScope.launch {
			if ( shouldResume && sessionIdentifier != null && sequenceNumber != null ) {
				sendResume( webSocket, sessionIdentifier, sequenceNumber )
				shouldResume = false
			} else {
				sendIdentify( webSocket )
			}
		}
	}

	// https://discord.com/developers/docs/topics/gateway-events#reconnect
	private fun handleReconnectEvent() {
		DiscordRelay.LOGGER.debug( "We need to reconnect!" )

		coroutineScope.launch {
			close( WebSocketCloseCode.GoingAway, "Told to reconnect." )

			DiscordRelay.LOGGER.debug( "Reconnecting as instructed..." )
			tryReconnect( true )
		}
	}

	private fun handleInvalidSessionEvent( shouldResume: Boolean ) {
		DiscordRelay.LOGGER.warn( "Our session is invalid!" )

		coroutineScope.launch {
			close( WebSocketCloseCode.GoingAway, "Session is invalid." )

			DiscordRelay.LOGGER.debug( "Reconnecting due to invalid session..." )
			tryReconnect( shouldResume )
		}
	}

	private fun handleHeartbeat( webSocket: WebSocket ) {
		val sequenceNumber = this.sequenceNumber
		DiscordRelay.LOGGER.debug( "Heartbeat requested at sequence number '$sequenceNumber'." )

		coroutineScope.launch {
			sendHeartbeat( webSocket, sequenceNumber )
		}
	}

	private fun handleHeartbeatAcknowledgement() {
		DiscordRelay.LOGGER.debug( "Heartbeat acknowledged." )
		heartbeatAcknowledgementConfirmation?.complete( Unit )
	}

	// https://discord.com/developers/docs/topics/gateway-events#ready
	private fun handleReadyEvent( data: Gateway.Event.Data.Ready ) {
		DiscordRelay.LOGGER.info( "Ready as '${ data.user.name }#${ data.user.discriminator }' / '@${ data.user.name }' (${ data.user.identifier })." )
		myIdentifier = data.user.identifier

		sessionIdentifier = data.sessionIdentifier
		resumeBaseUrl = data.resumeUrl
		DiscordRelay.LOGGER.debug( "Set session identifier to '$sessionIdentifier' & resume base URL to '$resumeBaseUrl'." )
	}

	// https://discord.com/developers/docs/topics/gateway-events#message-create
	private fun handleMessageCreate( message: Gateway.Event.Data.MessageCreate ) {
		DiscordRelay.LOGGER.debug( "Received message '${ message.content }' (${ message.identifier }) in channel ${ message.channelIdentifier } from '@${ message.author.name }' (${ message.author.identifier })." )

		if ( message.channelIdentifier != configuration.discord.channels.relay.identifier ) {
			DiscordRelay.LOGGER.debug( "Ignoring non-relay channel message (${ message.identifier }) from '@${ message.author.name }' (${ message.author.identifier })." )
			return
		}

		if ( message.content.isBlank() ) {
			DiscordRelay.LOGGER.debug( "Ignoring empty message (${ message.identifier }) from '@${ message.author.name }' (${ message.author.identifier })." )
			return
		}

		if ( message.author.isBot == true || message.author.isSystem == true ) {
			DiscordRelay.LOGGER.debug( "Ignoring bot/system message '${ message.content }' (${ message.identifier }) from '@${ message.author.name }' (${ message.author.identifier })." )
			return
		}

		if ( message.author.identifier == myIdentifier ) {
			DiscordRelay.LOGGER.debug( "Ignoring my message '${ message.content }' (${ message.identifier }) from '@${ message.author.name }' (${ message.author.identifier })." )
			return
		}

		DiscordRelay.LOGGER.info( "Relaying Discord message '${ message.content }' (${ message.identifier}) from '@${ message.author.name }' (${ message.author.identifier })..." )

		val memberRoleColor = getMemberRoleColor( message.member )
		val playerStyle = getStyleOrDefault( memberRoleColor )
		DiscordRelay.LOGGER.debug( "Member role color is '$memberRoleColor' & player style is '${ playerStyle.color?.rgb }'" )

		val chatMessage: Text = Text.literal( "" )
			.append( Text.literal( "(Discord) " ).setStyle( Style.EMPTY.withColor( TextColor.fromFormatting( Formatting.BLUE ) ) )
			.append( Text.literal( message.member?.displayName ?: message.author.name ).setStyle( playerStyle ) ) )
			.append( Text.literal( ": " ) )
			.append( Text.literal( message.content ) )

		playerManager.broadcast( chatMessage, false )
	}

	// https://discord.com/developers/docs/topics/gateway-events#guild-create
	private fun handleGuildCreate( guild: Gateway.Event.Data.GuildCreate ) {
		if ( guild.identifier != configuration.discord.server.identifier ) {
			DiscordRelay.LOGGER.debug( "Ignoring guild create event for guild '${ guild.name }' (${ guild.identifier })." )
			return
		}

		serverRoles = guild.roles.associateBy { it.identifier }
	}

	private fun getMemberRoleColor( member: Guild.Member? ): Int? = member?.roleIdentifiers
		?.map { serverRoles?.get( it ) ?: return null }
		?.maxByOrNull { it.position }
		?.color

	private fun getStyleOrDefault( rgb: Int?, formatting: Formatting = Formatting.GREEN ): Style =
		if ( rgb != null ) Style.EMPTY.withColor( rgb ) else Style.EMPTY.withColor( TextColor.fromFormatting( formatting ) )

	private fun processMessage( webSocket: WebSocket, message: String ) {
		val event = JSON.decodeFromString<Gateway.Event>( message )
		DiscordRelay.LOGGER.debug( "Received JSON payload '$message' from the WebSocket." )

		if ( event.sequenceNumber != null ) {
			DiscordRelay.LOGGER.debug( "Incremented sequence number from $sequenceNumber to ${ event.sequenceNumber }." )
			sequenceNumber = event.sequenceNumber
		}

		when ( event.operationCode ) {
			Gateway.Event.OperationCode.Hello -> {
				if ( event.data == null ) throw IllegalStateException( "Received Gateway hello operation '${ event.name }' without data" )
				handleHelloEvent( webSocket, JSON.decodeFromJsonElement<Gateway.Event.Data.Hello>( event.data ) )
			}
			Gateway.Event.OperationCode.Reconnect -> handleReconnectEvent()
			Gateway.Event.OperationCode.InvalidSession -> {
				if ( event.data == null ) throw IllegalStateException( "Received Gateway invalid session operation '${ event.name }' without data" )
				handleInvalidSessionEvent( JSON.decodeFromJsonElement<Boolean>( event.data ) )
			}
			Gateway.Event.OperationCode.Heartbeat -> handleHeartbeat( webSocket )
			Gateway.Event.OperationCode.HeartbeatAcknowledgement -> handleHeartbeatAcknowledgement()

			// https://discord.com/developers/docs/topics/gateway-events#receive-events
			Gateway.Event.OperationCode.Dispatch -> {
				if ( event.data == null ) throw IllegalStateException( "Received Gateway event '${ event.name }' without data" )

				when ( event.name ) {
					Gateway.Event.Name.Ready -> handleReadyEvent( JSON.decodeFromJsonElement<Gateway.Event.Data.Ready>( event.data ) )
					Gateway.Event.Name.MessageCreate -> handleMessageCreate( JSON.decodeFromJsonElement<Gateway.Event.Data.MessageCreate>( event.data ) )
					Gateway.Event.Name.GuildCreate -> handleGuildCreate( JSON.decodeFromJsonElement<Gateway.Event.Data.GuildCreate>( event.data ) )

					else -> DiscordRelay.LOGGER.debug( "Ignoring Gateway event '${ event.name }' with data '${ JSON.encodeToString( event.data ) }'." )
				}
			}

			else -> DiscordRelay.LOGGER.debug( "Ignoring Gateway operation ${ event.operationCode } with data '${ JSON.encodeToString( event.data ) }'." )
		}
	}

	private suspend fun tryReconnect( shouldResume: Boolean ) {
		val duration = ( 2.0.pow( reconnectCount ) * 1000 ).toLong()
		DiscordRelay.LOGGER.debug( "Connection attempt $reconnectCount, waiting $duration milliseconds before reconnecting..." )
		delay( duration )

		val resumeBaseUrl = resumeBaseUrl
		val baseUrl = if ( shouldResume && resumeBaseUrl != null ) resumeBaseUrl else API.getGateway().url
		DiscordRelay.LOGGER.debug( "Trying reconnect to '$baseUrl' (Resume: $shouldResume)..." )
		this.shouldResume = shouldResume
		open( baseUrl )

		reconnectCount++
		DiscordRelay.LOGGER.debug( "Incremented reconnection count to $reconnectCount." )
	}

	private fun cleanupAfterError() {
		DiscordRelay.LOGGER.debug( "Resetting state due to WebSocket error..." )

		this.webSocket = null

		this.sessionIdentifier = null
		this.resumeBaseUrl = null
		this.shouldResume = false

		this.connectionClosureConfirmation?.cancel()
		this.connectionClosureConfirmation = null

		this.heartbeatAcknowledgementConfirmation?.cancel()
		this.heartbeatAcknowledgementConfirmation = null

		this.heartbeatJob?.cancel()
		this.heartbeatJob = null

		this.myIdentifier = null
	}

	private inner class Listener: WebSocket.Listener {
		private val messageBuilder = StringBuilder()

		override fun onOpen( webSocket: WebSocket ) {
			DiscordRelay.LOGGER.debug( "WebSocket connection opened." )
			webSocket.request( 1 )
		}

		override fun onClose( webSocket: WebSocket, code: Int, reason: String? ): CompletionStage<*> {
			DiscordRelay.LOGGER.debug( "WebSocket connection closed with code $code & reason '$reason'." )

			DiscordRelay.LOGGER.debug( "Cancelling background heartbeating job..." )
			heartbeatJob?.cancel()

			DiscordRelay.LOGGER.debug( "Cancelling gateway coroutines..." )
			coroutineScope.cancel()

			// TODO: Check code against https://discord.com/developers/docs/topics/opcodes-and-status-codes#gateway-gateway-close-event-codes
			if ( code == WebSocketCloseCode.GoingAway || code == WebSocketCloseCode.ProtocolError ) {
				DiscordRelay.LOGGER.debug( "Reconnecting due to non-1000 close code..." )
				coroutineScope.launch {
					tryReconnect( false )
				}
			}

			return CompletableFuture.completedFuture( null )
		}

		override fun onText( webSocket: WebSocket, data: CharSequence?, isLastMessage: Boolean ): CompletionStage<*>? {
			DiscordRelay.LOGGER.debug( "Received text chunk of ${ data?.length } character(s)." )

			messageBuilder.append( data )

			if ( isLastMessage ) {
				processMessage( webSocket, messageBuilder.toString() )
				messageBuilder.clear()
			}

			webSocket.request( 1 )

			return CompletableFuture.completedFuture( null )
		}

		override fun onError( webSocket: WebSocket, error: Throwable? ) {
			DiscordRelay.LOGGER.error( "WebSocket error: '$error'" )

			// We're in a disconnected state but none of the closing code has run, so reset everything manually
			cleanupAfterError()

			DiscordRelay.LOGGER.debug( "Reconnecting due to WebSocket error..." )
			coroutineScope.launch {
				tryReconnect( false )
			}
		}
	}

	data class ClosureConfirmation(
		val closeCode: Int,
		val closeReason: String? = null,
		val isServerStopping: Boolean = false
	)
}
